File: frontend/src/api/session.ts
/**
 * Session API Client
 * Endpoints untuk managing driving sessions
 */

import axios from 'axios';
import { getAccessToken } from '../utils/auth';

const API_BASE_URL = 'http://localhost:8000/api/v1';

export interface SessionCreate {
    session_name: string;
    device_type?: string;
    calibration_data?: any;
    settings?: any;
}

export interface SessionUpdate {
    session_name?: string;
    session_status?: 'active' | 'completed' | 'failed';
    settings?: any;
    ended_at?: string;
    duration_seconds?: number;
    avg_fatigue_score?: number;
    max_fatigue_score?: number;
    alert_count?: number;
}

export interface ViolationData {
    type: string;
    points: number;
    timestamp: number;
    description: string;
}

export interface GameSessionData {
    routeName: string;
    totalWaypoints: number;
    reachedCount: number;
    missedCount: number;
    completionTime: number;
    violations: ViolationData[];
    totalViolationPoints: number;
    eegData?: any[];
    gameMetrics?: {
        averageSpeed: number;
        maxSpeed: number;
        collisions: number;
        laneDeviations: number;
        totalDistance: number;
    };
}

export interface SessionResponse {
    id: string;
    user_id: string;
    session_name: string;
    device_type?: string;
    session_status: string;
    started_at: string;
    ended_at?: string;
    duration_seconds?: number;
    avg_fatigue_score?: number;
    max_fatigue_score?: number;
    alert_count?: number;
    settings?: any;
}

export interface SessionListResponse {
    total: number;
    sessions: SessionResponse[];
    page: number;
    page_size: number;
}

export const sessionApi = {
    /**
     * Create a new session
     */
    create: async (data: SessionCreate): Promise<SessionResponse> => {
        const token = getAccessToken();
        if (!token) {
            console.error('❌ No access token found in localStorage');
            throw new Error('Not authenticated. Please log in first.');
        }

        // Debug: Show token info
        console.log('✅ Access token found, creating session...');
        console.log('📝 Token (first 50 chars):', token.substring(0, 50) + '...');
        console.log('📝 Session data:', data);

        try {
            const response = await axios.post(`${API_BASE_URL}/sessions`, data, {
                headers: {
                    'Authorization': `Bearer ${token}`
                }
            });
            console.log('✅ Session created successfully:', response.data);
            return response.data;
        } catch (error: any) {
            console.error('❌ Session creation failed');
            console.error('Error status:', error.response?.status);
            console.error('Error data:', error.response?.data);
            console.error('Error message:', error.message);
            throw error;
        }
    },

    /**
     * Get session by ID
     */
    get: async (sessionId: string): Promise<SessionResponse> => {
        const token = getAccessToken();
        if (!token) {
            throw new Error('Not authenticated. Please log in first.');
        }
        const response = await axios.get(`${API_BASE_URL}/sessions/${sessionId}`, {
            headers: {
                'Authorization': `Bearer ${token}`
            }
        });
        return response.data;
    },

    /**
     * End a session
     */
    end: async (sessionId: string): Promise<SessionResponse> => {
        const token = getAccessToken();
        if (!token) {
            throw new Error('Not authenticated. Please log in first.');
        }
        const response = await axios.patch(
            `${API_BASE_URL}/sessions/${sessionId}/end`,
            {},
            {
                headers: {
                    'Authorization': `Bearer ${token}`
                }
            }
        );
        return response.data;
    },

    /**
     * Complete a session with final data
     */
    complete: async (sessionId: string, data?: SessionUpdate): Promise<SessionResponse> => {
        const token = getAccessToken();
        if (!token) {
            throw new Error('Not authenticated. Please log in first.');
        }
        
        // First update session with final stats if provided
        if (data) {
            await axios.patch(
                `${API_BASE_URL}/sessions/${sessionId}`,
                data,
                {
                    headers: {
                        'Authorization': `Bearer ${token}`
                    }
                }
            );
        }
        
        // Then mark as complete
        const response = await axios.post(
            `${API_BASE_URL}/sessions/${sessionId}/complete`,
            {},
            {
                headers: {
                    'Authorization': `Bearer ${token}`
                }
            }
        );
        return response.data;
    },

    /**
     * Update a session
     */
    update: async (sessionId: string, data: SessionUpdate): Promise<SessionResponse> => {
        const token = getAccessToken();
        if (!token) {
            throw new Error('Not authenticated. Please log in first.');
        }
        const response = await axios.patch(
            `${API_BASE_URL}/sessions/${sessionId}`,
            data,
            {
                headers: {
                    'Authorization': `Bearer ${token}`
                }
            }
        );
        return response.data;
    },

    /**
     * List user's sessions with pagination
     */
    list: async (page: number = 1, pageSize: number = 20, status?: string): Promise<SessionListResponse> => {
        const token = getAccessToken();
        if (!token) {
            throw new Error('Not authenticated. Please log in first.');
        }
        
        const params = new URLSearchParams({
            page: page.toString(),
            page_size: pageSize.toString(),
        });
        if (status) {
            params.append('status', status);
        }
        
        const response = await axios.get(
            `${API_BASE_URL}/sessions?${params.toString()}`,
            {
                headers: {
                    'Authorization': `Bearer ${token}`
                }
            }
        );
        return response.data;
    },

    /**
     * Delete a session
     */
    delete: async (sessionId: string): Promise<void> => {
        const token = getAccessToken();
        if (!token) {
            throw new Error('Not authenticated. Please log in first.');
        }
        await axios.delete(
            `${API_BASE_URL}/sessions/${sessionId}`,
            {
                headers: {
                    'Authorization': `Bearer ${token}`
                }
            }
        );
    }
};

File: backend/app/schemas/session.py
"""
Session Schemas
Pydantic models for driving session data validation
Week 2, Tuesday - API Routes
"""

from pydantic import BaseModel, Field
from typing import Optional, Dict, Any
from uuid import UUID
from datetime import datetime


class SessionBase(BaseModel):
    """Base session schema"""
    session_name: str = Field(..., min_length=1, max_length=255)
    device_type: Optional[str] = Field(None, max_length=100, description="EEG device type (e.g., Muse 2)")
    settings: Optional[Dict[str, Any]] = Field(default_factory=dict, description="Game settings and thresholds")


class SessionCreate(SessionBase):
    """Schema for creating a new session"""
    calibration_data: Optional[Dict[str, Any]] = Field(default_factory=dict, description="Baseline calibration data")


class SessionUpdate(BaseModel):
    """Schema for updating session"""
    session_name: Optional[str] = Field(None, min_length=1, max_length=255)
    session_status: Optional[str] = Field(None, pattern="^(active|completed|failed)$")
    calibration_data: Optional[Dict[str, Any]] = None
    settings: Optional[Dict[str, Any]] = None
    ended_at: Optional[datetime] = None
    duration_seconds: Optional[int] = None
    avg_fatigue_score: Optional[float] = Field(None, ge=0, le=100)
    max_fatigue_score: Optional[float] = Field(None, ge=0, le=100)
    alert_count: Optional[int] = Field(None, ge=0)


class SessionResponse(BaseModel):
    """Schema for session response"""
    id: UUID
    user_id: UUID
    session_name: str
    device_type: Optional[str] = None
    session_status: str
    calibration_data: Optional[Dict[str, Any]] = None
    settings: Optional[Dict[str, Any]] = None
    started_at: datetime
    ended_at: Optional[datetime] = None
    duration_seconds: Optional[int] = None
    avg_fatigue_score: Optional[float] = None
    max_fatigue_score: Optional[float] = None
    alert_count: int
    
    model_config = {"from_attributes": True}


class SessionListResponse(BaseModel):
    """Schema for paginated session list"""
    total: int
    sessions: list[SessionResponse]
    page: int = 1
    page_size: int = 20
